[IMPROVED] Replace panic with error during configuration parsing/reload

matthiashanel opened the pull request:
This will also result in an error with line number close to the issue.

Basically move parsing into own function.
In there, recover and turn panic into an error.

kozlovic commented:
@matthiashanel Let's apply the `go fmt` and then you can: `git commit -s -a --amend` to rewrite this commit, followed by a `git push --force` to overwrite this wrong commit.

kozlovic commented on server/configs/panic.conf:
```diff
@@ -0,0 +1 @@
+port "thisstringtripsapanic"
```
New line?

matthiashanel commented:
I added my change to maps as well. 
To keep the entire change smaller, I switched back to *[]error and use of a last token pointer.
(could have used tk as well, but then the change would have been bigger as I need a single variable and tk is shadowed at times. 
I am not sure what the license issues are about

derekcollison commented:
license issues are separate thing and it FOSSA changing the way it looks at projects and now inspects all the tech used to build the server :/

kozlovic commented on server/opts_test.go:
```diff
@@ -1142,6 +1142,17 @@ func TestMalformedClusterAddress(t *testing.T) {
 }
 
+func TestPanic(t *testing.T) {
+	opts, err := ProcessConfigFile("./configs/panic.conf")
```
Sorry, I should have mentioned that in the first review. We try now to not add static config files for tests when possible. We have some utilities that will create temporary config files, etc.. So you could rewrite the test this way (and therefore get rid of the new `./configs/panic.conf` file):
```
conf := createConfFile(t, []byte(`port: "this_string_trips_a_panic"`))
defer os.Remove(conf)
opts, err := ProcessConfigFile(conf)
if err == nil {
    t.Fatalf("Expected an error reading config file: got %+v", opts)
} else {
    if !strings.Contains(err.Error(), ":1:0: interface conversion:") {
        t.Fatalf("This was supposed to trip a panic on interface conversion right at the beginning")
    }
}
```

matthiashanel replied:
incorporated that

matthiashanel commented on server/opts.go:
```diff
@@ -437,358 +461,368 @@ func (o *Options) ProcessConfigFile(configFile string) error {
 	}
 
+	var lt token
+	defer convertPanicToErrorList(&lt, &errors)
```
removed

matthiashanel commented on server/opts.go:
```diff
@@ -373,20 +373,44 @@ type token interface {
-func configureSystemAccount(o *Options, m map[string]interface{}) error {
+func configureSystemAccount(o *Options, m map[string]interface{}) (err error) {
+	var lt token
+	defer convertPanicToError(&lt, &err)
```
left that one in there to stay with the same pattern

matthiashanel commented on server/opts.go:
```diff
@@ -373,20 +373,44 @@ type token interface {
+
+// use in defer to recover from panic and turn it into an error associated with last token
+func convertPanicToError(lastToken *token, e *error) {
+	if e != nil && *e != nil {
```
fixed that test

matthiashanel commented on server/opts.go:
```diff
@@ -373,20 +373,44 @@ type token interface {
 
+// use in defer to recover from panic and turn it into an error associated with last token
+func convertPanicToErrorList(lastToken *token, errors *[]error) {
+	if lastToken != nil && *lastToken != nil {
```
lastToken is a *token as this code is within a dedicated function.
made a change to create a regular error if no lastToken was present.

kozlovic commented on server/opts.go:
```diff
@@ -373,20 +373,54 @@ type token interface {
+	} else if lastToken != nil && *lastToken != nil {
+		*errors = append(*errors, &configErr{*lastToken, fmt.Sprint(err)})
+	} else {
+		*errors = append(*errors, fmt.Errorf("Encountered panic without a token %v", err))
```
See error reported by staticcheck. Don't use capital for errors.

matthiashanel replied:
changed

kozlovic commented on server/opts.go:
```diff
@@ -373,20 +373,54 @@ type token interface {
+// use in defer to recover from panic and turn it into an error associated with last token
+func convertPanicToError(lastToken *token, e *error) {
+	// only recover if an error can be stored
+	if e == nil || *e != nil {
```
No action, but I got confused for a second here (because of the comment). Yes, we need `e` to be not nil so that we can store the constructed error (out of the panic) but we don't need to do anything if `*e` is not nil, which means that this function was invoked with an error (and likely not a panic).

kozlovic commented on server/opts.go:
```diff
@@ -373,20 +373,54 @@ type token interface {
 // configureSystemAccount configures a system account
 // if present in the configuration.
-func configureSystemAccount(o *Options, m map[string]interface{}) error {
+func configureSystemAccount(o *Options, m map[string]interface{}) (err error) {
```
Personal preference, but I would have called this one `retErr error` (and obviously pass that to convertPanicToError() down below), to indicate that this is the returned error and avoid possible shadowing in the rest of the function.

matthiashanel replied:
changed
